agent/lib/kontena/service_pods/lifecycle_hook_manager.rb

```diff
@@ -0,0 +1,159 @@
+          begin
+            info "running pre_start hook: #{hook['cmd']}"
+            service_config = config_container(service_pod.dup, hook['cmd'])
+            service_config['name'] = "#{service_config['name']}-#{SecureRandom.urlsafe_base64(5)}"
+            service_container = create_container(service_config)
+            log_hook_output(service_container.id, "running pre_start hook: #{hook['cmd']}", 'stdout')
+            service_container.tap(&:start).attach
+            if service_container.state['ExitCode'] != 0
+              raise "Failed to execute pre_start hook: #{hook['cmd']}, exit code: #{service_container.state['ExitCode']}"
+            end
+          ensure
+            cleanup_container(service_container) if service_container
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        raise exc
+      end
+
+      # @param [Docker::Container] service_container
+      # @return [Boolean]
+      def on_post_start(service_container)
+        hooks_for('post_start').each do |hook|
+          info "running post_start hook: #{hook['cmd']}"
+          command = build_cmd(hook['cmd'])
+          log_hook_output(service_container.id, "running post_start hook: #{hook['cmd']}", 'stdout')
+          _, _, exit_code = service_container.exec(command) { |stream, chunk|
+            log_hook_output(service_container.id, chunk, stream)
+          }
+          if exit_code != 0
+            raise "Failed to execute post_start hook: #{hook['cmd']} (exit code: #{exit_code}"
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        error exc.message
+        false
+      end
+
+      # @param [Docker::Container] service_container
+      # @return [Boolean]
+      def on_pre_stop(service_container)
+        hooks_for('pre_stop').each do |hook|
+          info "running pre_stop hook: #{hook['cmd']}"
+          command = build_cmd(hook['cmd'])
+          log_hook_output(service_container.id, "running pre_stop hook: #{hook['cmd']}", 'stdout')
+          _, _, exit_code = service_container.exec(command) { |stream, chunk|
+            log_hook_output(service_container.id, chunk, stream)
+          }
+          if exit_code != 0
+            raise "Failed to execute pre_stop hook: #{hook['cmd']} (exit code: #{exit_code}"
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        error exc.message
+        false
+      end
+
+      # @param [String] type
+      # @return [Array<Hash>]
+      def hooks_for(type)
+        hooks = service_pod.hooks.select { |h|
+          h['type'] == type && !h['oneshot'] || (h['oneshot'] && !oneshot_cache.include?(h))
+        }
+        hooks.each do |h|
+          if h['oneshot'] && !oneshot_cache.include?(h)
```

Comment:
This `oneshot_cache.include?` is subtly broken, because it's using the `{ 'type' => ..., 'cmd' => ..., 'oneshot' => true/false}` as a "unique" key for the `oneshot_cache`, but it isn't: the sequence of hooks for a type can contain the same `cmd` multiple times. With some `DEBUG` logging inside this `if` it's apparent that this is only marking the first occurence of the `sleep 5` command, and it's not marking the second occurence:

```
D, [2017-09-20T12:27:22.021211 #1] DEBUG -- Kontena::ServicePods::Creator: creating container: hooks-oneshot.redis-1
D, [2017-09-20T12:27:22.311482 #1] DEBUG -- Kontena::ServicePods::Creator: container created: hooks-oneshot.redis-1
I, [2017-09-20T12:27:23.338078 #1]  INFO -- Kontena::ServicePods::Creator: service started: hooks-oneshot/redis-1
D, [2017-09-20T12:27:23.367677 #1] DEBUG -- Kontena::ServicePods::LifecycleHookManager: marking oneshot hook: {"type"=>"post_start", "cmd"=>"date && echo \"pre-sleep\"", "oneshot"=>true}
D, [2017-09-20T12:27:23.477041 #1] DEBUG -- Kontena::ServicePods::LifecycleHookManager: marking oneshot hook: {"type"=>"post_start", "cmd"=>"sleep 5", "oneshot"=>true}
D, [2017-09-20T12:27:23.487460 #1] DEBUG -- Kontena::ServicePods::LifecycleHookManager: marking oneshot hook: {"type"=>"post_start", "cmd"=>"date && echo \"mid-sleep\"", "oneshot"=>true}
D, [2017-09-20T12:27:23.496961 #1] DEBUG -- Kontena::ServicePods::LifecycleHookManager: marking oneshot hook: {"type"=>"post_start", "cmd"=>"date && echo \"post-sleep\"", "oneshot"=>true}
I, [2017-09-20T12:27:23.509313 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "first hook (non-oneshot)"
I, [2017-09-20T12:27:23.854263 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "pre-sleep"
I, [2017-09-20T12:27:24.102533 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: sleep 5
I, [2017-09-20T12:27:29.284095 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "mid-sleep"
I, [2017-09-20T12:27:29.448416 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: sleep 5
I, [2017-09-20T12:27:34.595113 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "post-sleep"
I, [2017-09-20T12:27:34.769682 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "last hook (non-oneshot)"
```

I suggest that the agent `oneshot_cache` would also use `h['id']` as the key. It's feels very wrong to have overlapping keys for distinct entries in the `oneshot_cache`.

It's also subtly non-broken for the `service_pod.hooks.select { |h| ... oneshot_cache.include?(h)  }` case, because this does all of the per-type `oneshot_cache.include?(h)` lookups before marking any of them.  There's a risk of it skipping the duplicate hooks if that were written any differently.